tchannel/tornado/connection.py

```diff
@@ -664,7 +663,7 @@ def keep_reading(f):
                 lambda f: io_loop.spawn_callback(self.fill),
             )
 
-        io_loop.add_future(read_message(self.io_stream), keep_reading)
```

Comment:
Explained in the description but basically the `add_future` is the reason that the `read_message` future does not get consumed.

    io_loop.add_future(f, cb) == io_loop.add_callback(lambda: f.add_done_callback(cb))

If the IOLoop is stopped before the next iteration, it never calls `add_done_callback` and `f` stays unconsumed. 

Why the IOLoop is getting stopped before then, I'm not entirely sure, but I believe it might be due to timeouts causing a higher level Future to finish early.